phylogeny/upgma.cpp

```diff
@@ -0,0 +1,63 @@
+#include "upgma.hpp"
+
+#include <unordered_set>
+#include <vector>
+
+namespace ts {
+
+Tree upgma(const std::vector<std::vector<double>> &dist_mat) {
+    if (dist_mat.empty()) {
+        return {};
+    }
+
+    Tree result(2 * dist_mat.size() - 1);
+    // {nodeId, nodeCount} pairs of all the cluster roots
+    std::unordered_map<uint32_t, uint32_t> roots;
+    std::unordered_map<uint32_t, std::unordered_map<uint32_t, double>> D;
+    for (uint32_t i = 0; i < dist_mat.size(); ++i) {
+        roots.insert({ i, 1 });
+        result[i] = { 0, NO_CHILD, NO_CHILD };
+        for (uint32_t j = 0; j < dist_mat.size(); ++j) {
+            D[i][j] = dist_mat[i][j];
+        }
+    }
+    for (uint32_t step = 0; step < dist_mat.size() - 1; ++step) {
+        double minDist = std::numeric_limits<double>::max();
+        uint32_t min_i, min_j;
+        for (const auto &root1 : roots) {
+            for (const auto &root2 : roots) {
```

Comment:
while not very efficient, it does the job. If it's too slow we could at some point replace the nested unordered map by a single unordered map or vector<vector<>>.
Also, we could keep a heap of smallest distances, to prevent this double for loop.